pkg/controllers/certificate/cert_rotation_controller.go

```diff
@@ -0,0 +1,317 @@
+	}
+
+	return controllerruntime.Result{RequeueAfter: c.CertRotationCheckingInterval}, nil
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *CertRotationController) SetupWithManager(mgr controllerruntime.Manager) error {
+	return controllerruntime.NewControllerManagedBy(mgr).
+		For(&clusterv1alpha1.Cluster{}, builder.WithPredicates(c.PredicateFunc)).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiterflag.DefaultControllerRateLimiter(c.RatelimiterOptions),
+		}).
+		Complete(c)
+}
+
+func (c *CertRotationController) syncCertRotation(secret *corev1.Secret) error {
+	karmadaKubeconfig, err := getKubeconfigFromSecret(secret)
+	if err != nil {
+		return err
+	}
+
+	clusterName := karmadaKubeconfig.Contexts[karmadaKubeconfig.CurrentContext].AuthInfo
+	if clusterName == "" {
+		return fmt.Errorf("failed to get cluster name, the current context is %s", karmadaKubeconfig.CurrentContext)
+	}
+
+	oldCertData := karmadaKubeconfig.AuthInfos[clusterName].ClientCertificateData
+
+	shouldRotate, err := c.shouldRotateCert(oldCertData)
+	if err != nil {
+		return err
+	}
+	if !shouldRotate {
+		return nil
+	}
+
+	oldCert, err := certutil.ParseCertsPEM(oldCertData)
+	if err != nil {
+		return fmt.Errorf("unable to parse old certificate: %v", err)
+	}
+
+	// create a new private key
+	keyData, err := keyutil.MakeEllipticPrivateKeyPEM()
+	if err != nil {
+		return err
+	}
+
+	privateKey, err := keyutil.ParsePrivateKeyPEM(keyData)
+	if err != nil {
+		return fmt.Errorf("invalid private key for certificate request: %v", err)
+	}
+
+	csr, err := c.createCSRInControlPlane(clusterName, privateKey, oldCert)
+	if err != nil {
+		return fmt.Errorf("failed to create csr in control plane, err is: %v", err)
+	}
+
+	var newCertData []byte
+	klog.V(1).Infof("Waiting for the client certificate to be issued")
+	err = wait.Poll(1*time.Second, 5*time.Minute, func() (done bool, err error) {
+		csr, err := c.KubeClient.CertificatesV1().CertificateSigningRequests().Get(context.TODO(), csr, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("failed to get the cluster csr %s. err: %v", clusterName, err)
+		}
+
+		if csr.Status.Certificate != nil {
+			klog.V(1).Infof("Signing certificate successfully")
+			newCertData = csr.Status.Certificate
+			return true, nil
+		}
+
+		klog.V(1).Infof("Waiting for the client certificate to be issued")
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	karmadaKubeconfig.AuthInfos[clusterName].ClientCertificateData = newCertData
+	karmadaKubeconfig.AuthInfos[clusterName].ClientKeyData = keyData
+
+	karmadaKubeconfigBytes, err := clientcmd.Write(*karmadaKubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to serialize karmada-agent kubeConfig. %v", err)
+	}
+
+	secret.Data["karmada-kubeconfig"] = karmadaKubeconfigBytes
+	secret.ResourceVersion = ""
```

Comment:
Just a question, why empty resource version here?
I guess we can add a comment here.

Reply:
I guess it is because the `CreateOrUpdateSecret` not works as expected in case of the secret already exists:
https://github.com/karmada-io/karmada/blob/a926ca4f4960154656fcc4c4277ecaf1f4dd8995/pkg/util/secret.go#L87-L99

Right?
If yes, I think we need to fix it.

Reply:
Yes

```
E1026 09:19:42.142011       1 cert_rotation_controller.go:106] Failed to rotate the certificate of karmada-agent for the given member cluster: member68, err is: unable to create secret, err: resourceVersion should not be set on objects to be created
```
I will fix it.